torchmultimodal/models/mdetr.py

```diff
@@ -0,0 +1,530 @@
+            backbone: torch module of the backbone to be used. See backbone.py
+            text_encoder: torch module of text encoder to be used
+            transformer: torch module of the transformer architecture. See transformer.py
+
+        """
+        super().__init__()
+        self.image_backbone = image_backbone
+        self.text_encoder = text_encoder
+        self.resizer = resizer
+        self.transformer = transformer
+        self.pos_embed = pos_embed
+        self.input_proj = input_proj
+        self.class_embed = class_embed
+        self.bbox_embed = bbox_embed
+        self.query_embed = query_embed
+
+    def forward(self, images: NestedTensor, text: Tensor, text_attention_mask: Tensor):
+        """The forward expects a NestedTensor of images, which consists of:
+           - images.tensor: batched images, of shape [batch_size x 3 x H x W]
+           - images.mask: a binary mask of shape [batch_size x H x W], containing 1 on padded pixels
+        It also expects a tensor of texts and a tensor of text attention masks.
+
+        It returns a dict with the following elements:
+           - "pred_logits": the classification logits (including no-object) for all queries.
+                            Shape= [batch_size x num_queries x (num_classes + 1)]
+           - "pred_boxes": The normalized boxes coordinates for all queries, represented as
+                           (center_x, center_y, height, width). These values are normalized in [0, 1],
+                           relative to the size of each individual image (disregarding possible padding).
+                           See PostProcess for information on how to retrieve the unnormalized bounding box.
+        """
+
+        encoded_text = self.text_encoder(text, text_attention_mask)
+
+        # Transpose memory because pytorch's attention expects sequence first
+        text_memory = encoded_text.transpose(0, 1)
+
+        # Invert attention mask that we get from huggingface because its the opposite in pytorch transformer
+        text_attention_mask = text_attention_mask.ne(1).bool()
+
+        features = self.image_backbone(images)
+        pos = [self.pos_embed(x).to(x.tensors.dtype) for x in features]
+
+        src, mask = features[-1].decompose()
+        query_embed = self.query_embed.weight
+
+        text_memory_resized = self.resizer(text_memory)
+
+        hs = self.transformer(
+            self.input_proj(src),
+            mask,
+            query_embed,
+            pos[-1],
+            text_memory=text_memory_resized,
+            img_memory=None,
+            text_attention_mask=text_attention_mask,
+        )
+        out = {}
+        outputs_class = self.class_embed(hs)
+        outputs_coord = self.bbox_embed(hs).sigmoid()
+
+        out.update(
+            {
+                "pred_logits": outputs_class[-1],
+                "pred_boxes": outputs_coord[-1],
+            }
+        )
+
+        return out
+
+
+class Transformer(nn.Module):
```

Comment:
is this straight up from original code?

Reply:
No, there are some changes. The main ones are factoring out the intermediate caching of encoder results, removing the text encoder, and cleaning up some unnecessary params. There are some other minor ones too. But for the most part the lower-level components are left intact